Contest_API.md

```diff
@@ -751,8 +751,22 @@ Returned data:
 
 ### Groups
 
-Grouping of teams. At the World Finals these are the super regions, at
-regionals these are often different sites.
+Grouping of teams. At the World Finals these are the super regions; at other contests these
+may be the different sites, divisions, or types of contestants.
+
+Teams may belong to multiple groups. For instance, there may be a group for each site, a group for
+university teams, a group for corporate teams, and a group for ICPC-eligible teams. Teams could
+belong to two or three of these.
+When there are different kinds of groups for different purposes (e.g. sites vs divisions), each
+group or set of groups should have a different type attribute
+(e.g. `"type":"site"` and `"type":"division"`).
```

Comment:
I think we've discussed this before, but don't we want to require (or at least recommend) that for each group type, a team can at most (or exactly) be in one of those?

Reply:
I added a recommendation.

Reply:
> don't we want to require (or at least recommend) that for each group type, a team can at most (or exactly) be in one of those?

We should **absolutely** not require that for all group types. We could possibly add additional requirements for certain predefined types. E.g. it could make sense that if you are using groups of type "site" then every team has to be in exactly one site group.

> I added a recommendation.

I don't like that recommendation.

Reply:
Ok, I removed it. I did move a waffly "generally should not" to the site description - IMHO maybe this should be stronger, but I'd rather get to a point where people are happy to merge and debate tweaks on a call.

Reply:
> > don't we want to require (or at least recommend) that for each group type, a team can at most (or exactly) be in one of those?
> 
> We should **absolutely** not require that for all group types. We could possibly add additional requirements for certain predefined types. E.g. it could make sense that if you are using groups of type "site" then every team has to be in exactly one site group.

When does it make sense for a team to be part of multiple groups of the same group type? (Maybe I forgot a case we discussed before, but if anything I guess it would be useful to document here.)

Reply:
Well, since groups and types could be anything, the answer very much depends on how you use them.

here are some cases from the top of my head:
* If there are groups for "has a graduate program" and "offers doctorates degrees", a school could be in one, both or neither.
* When you have filtering rules such as "ICPC eligible" as we do at NCPC, you end up with a group for all and a group for only eligible, but you typically don't have one for only non-eligible. Assuming you further filter on e.g. country (again, as we do at NCPC), then the country level ICPC and non-ICPC groups would (maybe?) be of the same type, but all eligible teams would be in both.
* If you have a country scoreboard but the semantics is not "country of the university" (maybe because the teams do not represent universities) but rather "nationality of the team members", then a team with team members of different nationalities would be in several of these groups.

The actual point though is that we have not in any way specified exactly how groups are to be used, so we definitely can't assume that the groups of a certain type has properties such as non-overlapping or completeness (i.e. every team is in one group of the type). Again, it would be a good idea to add these guarantees to predefined specific types when that makes sense. E.g. it *does* make sense to say that site is a proper partition of the teams.

Reply:
Ok, but in your examples I don't see groups that could overlap *and* would naturally have the same type. So I think it's a reasonable requirement that a team can be part of _at most_ one group of each type. Thus a type is not yet a strict partition, but at least if you'd add a "other" group of that type it would be.

I think this requirement would be useful, for example for a tool that would want to display the "site" of a team. If you know that a team can be assigned at most one "site" group, then you can render that in a nicer way than if it could belong to multiple sites.

Reply:
> Ok, but in your examples I don't see groups that could overlap _and_ would naturally have the same type.

The first and third are exactly that. In the third example we would have groups for every country (with participants in the contest). Every team with a Swedish team member would be in the "Sweden" group, every team with a Canadian team member would be in the "Canada" group. So, if me and Tim make a team, our team would be in both those groups. Are you saying that these country groups would not be of the same type?

> I think this requirement would be useful, for example for a tool that would want to display the "site" of a team. If you know that a team can be assigned at most one "site" group, then you can render that in a nicer way than if it could belong to multiple sites.

Right, which is why I **do** think it would be very useful to say that groups of type `site` must be a partition of all teams.

Reply:
> > Ok, but in your examples I don't see groups that could overlap _and_ would naturally have the same type.
> 
> The first and third are exactly that. In the third example we would have groups for every country (with participants in the contest). Every team with a Swedish team member would be in the "Sweden" group, every team with a Canadian team member would be in the "Canada" group. So, if me and Tim make a team, our team would be in both those groups. Are you saying that these country groups would not be of the same type?

On the first, I don't really see why these need to have the same type, even have to have a type at all. For countries, I agree that's a possible use case. But then, will you also display multiple country flags on the scoreboard for that team?

> > I think this requirement would be useful, for example for a tool that would want to display the "site" of a team. If you know that a team can be assigned at most one "site" group, then you can render that in a nicer way than if it could belong to multiple sites.
> 
> Right, which is why I **do** think it would be very useful to say that groups of type `site` must be a partition of all teams.

I don't like it so much to attach special meaning to a particular group type `site`: that this one must be a partition of all teams. What about the following: next to `type` we introduce an attribute `label`. Then `type` should be used for groups that partition all teams, while `label` has no such restrictions.

Reply:
> On the first, I don't really see why these need to have the same type, even have to have a type at all.

Sure. If we would specify that types must be proper partition we could just set these groups up as type-less.

> For countries, I agree that's a possible use case. But then, will you also display multiple country flags on the scoreboard for that team?

Yes? I'm not planning to implement it any time soon, this is all hypothetical, but yes, that's how I would do it. Also, it's not *completely* made up, we actually had a scoreboard like that some 15-20 years ago, for a contest with teams of 3 that were just freely formed. and not associated with universities or such.

> What about the following: next to `type` we introduce an attribute `label`. Then `type` should be used for groups that partition all teams, while `label` has no such restrictions.

I **really** don't like to have two types of types with different semantics. Also, the name label to me implies some display property rather than a grouping property.

Reply:
I agree that `label` is maybe not good name and even useful enough to add. But I guess my point is: what is the value of `type` if it doesn't have some restrictions imposed on the groups that are part of a single type?
And I don't see a point in assigning a type to each group: if there's no clear reason to use `type`, I think it's perfectly fine to not assign a `type`, like with "has a graduate program" and "offers doctorates degrees".